src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs

```diff
@@ -45,6 +45,7 @@ internal class HttpInListener : ListenerHandler
         private readonly PropertyFetcher<HttpContext> startContextFetcher = new("HttpContext");
         private readonly PropertyFetcher<HttpContext> stopContextFetcher = new("HttpContext");
         private readonly PropertyFetcher<Exception> stopExceptionFetcher = new("Exception");
+        private readonly PropertyFetcher<Exception> stopExceptionFilterFetcher = new("ExceptionContext.Exception");
```

Comment:
Nice, thanks for discovering this @al-mac!

You got me curious, so I looked and found that we were indeed missing the events that are written before and after exception filters are invoked:

Before:
https://github.com/dotnet/aspnetcore/blob/36df8e5dc487e27cb85674ef3f144dca07f05c20/src/Mvc/Mvc.Core/src/MvcCoreDiagnosticListenerExtensions.cs#L396-L402

After:
https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Mvc/Mvc.Core/src/MvcCoreDiagnosticListenerExtensions.cs#L422-L434

I suspect that both of these events will cause our instrumentation to be invoked potentially duplicating the work we do in `OnException`.

Would you be able to write a test to validate this behavior?

We have some test apps for testing the ASP.NET Core instrumentation. I think the relevant place to configure an exception filter is here:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/720cbdd2f8172569861e23783d8a06b6d5950f30/test/TestApp.AspNetCore.6.0/Startup.cs#L38

And here's an example of a test that exercises the test app:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/720cbdd2f8172569861e23783d8a06b6d5950f30/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs#L156-L166

We might need a way to conditionally set up an exception filter in the test app for the purposes of testing your work in this PR. If you need some help tooling this into our test infrastructure, I can help out more next week.

Reply:
> It's possible that something else is going on. Are you sure OnException is not also getting invoked with a payload with an Exception property?

My memory clicked in and I think I know what the problem is. I'm nearly certain that `OnException` is invoked both for Exception filters _and_ the regular exception path. `PropertyFetcher` effectively caches the payload type it receives. All subsequent calls to the `PropertyFetcher` expect that type.  The problem is the first time `OnException` is invoked it is invoked due to an Exception filter.

We're not interested in the exception from exception filters because of the concerns I raised before (i.e., there may be multiple filters and they may handle the exception).

@CodeBlanch ran into this same issue in #2698.

The alternative he considered was maintaining a map of inner `PropertyFetcher`s based on type, but that has downsides for performance. Though in this case, since I don't think we're interested in the Exception filter path, I think we could modify `PropertyFetcher` to ignore it. If the payload from the events raised by Exception filters do not have a property named `Exception` then I think we should prevent the PropertyFetcher from caching anything.

I still think the first step is to write a test that uses Exception filters since this is a use case that affects the behavior of the instrumentation.

Reply:
Hey, i'm working on the stuff you asked and came to the same conclusion! You hinted that the exception should be handled regardless of the `ExceptionFilter` being there or not. I've investigated it and found that the reason for this behaviour was the `innerFetcher` was being populated with the generic one (The one that always returns `null` on any fetch) before the actual object that has the exception property was handled in the pipeline. But when it's their turn to be handled, the `innerFetcher` is not `null` anymore, so the correct `TypedPropertyFetch<,>` is never being constructed.

Another way to fix it is to change make the `CreateFetcherFromProperty` method return `null` when this happens, and to return `default` in the `TryFetch` method if the innerFetcher is still `null` after trying to create it. I still didn't validate if this will make other parts of the code break, but I'm working on it right now.

Reply:
Hi! I've changed the code to reflect what I described above and wrote a few tests to validate this behaviour.